File: RxNetWork-Java/RxNetWorkLibrary/src/main/java/io/reactivex/network/bus/RxBus.java
package io.reactivex.network.bus;

import android.support.annotation.NonNull;
import android.support.v4.util.ArrayMap;

import java.util.Map;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.network.util.RxUtils;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

/**
 * by y on 2017/2/22.
 */
public class RxBus {
    private final ArrayMap<Object, RxBusEvent> rxBusEventArrayMap;

    private RxBus() {
        rxBusEventArrayMap = new ArrayMap<>();
    }

    public static RxBus getInstance() {
        return RxBusHolder.rxBus;
    }

    private static class RxBusHolder {
        private static final RxBus rxBus = new RxBus();
    }

    public boolean post(@NonNull Object obj) {
        return post(obj, obj);
    }

    public boolean post(@NonNull Object tag, @NonNull Object obj) {
        RxBusEvent rxBusEvent = rxBusEventArrayMap.get(tag);
        if (!RxUtils.isEmpty(rxBusEvent)) {
            rxBusEvent.subject.onNext(obj);
            return true;
        }
        return false;
    }

    /**
     * 接受消息
     *
     * @param tag      标志
     * @param callBack 回调
     */
    public <T> DisposableObserver register(@NonNull final Object tag,
                                           @NonNull final RxBusCallBack<T> callBack) {
        RxBusEvent rxBusEvent = rxBusEventArrayMap.get(tag);
        if (RxUtils.isEmpty(rxBusEvent)) {
            rxBusEvent = new RxBusEvent();
            rxBusEvent.subject = PublishSubject.create().toSerialized();
            rxBusEvent.disposable =
                    rxBusEvent.subject
                            .ofType(callBack.busOfType())
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeWith(new RxBusObserver<T>() {
                                @Override
                                public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                                    super.onError(e);
                                    callBack.onBusError(e);
                                }

                                @Override
                                public void onNext(@io.reactivex.annotations.NonNull T t) {
                                    super.onNext(t);
                                    callBack.onBusNext(t);
                                }
                            });
            rxBusEventArrayMap.put(tag, rxBusEvent);
        }
        return rxBusEvent.disposable;
    }


    /**
     * 接受消息
     *
     * @param tag      标志
     * @param callBack 回调
     */
    public <T> DisposableObserver registerNoThread(@NonNull final Object tag,
                                                   @NonNull final RxBusCallBack<T> callBack) {
        RxBusEvent rxBusEvent = rxBusEventArrayMap.get(tag);
        if (RxUtils.isEmpty(rxBusEvent)) {
            rxBusEvent = new RxBusEvent();
            rxBusEvent.subject = PublishSubject.create().toSerialized();
            rxBusEvent.disposable =
                    rxBusEvent.subject
                            .ofType(callBack.busOfType())
                            .subscribeWith(new RxBusObserver<T>() {
                                @Override
                                public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                                    super.onError(e);
                                    callBack.onBusError(e);
                                }

                                @Override
                                public void onNext(@io.reactivex.annotations.NonNull T t) {
                                    super.onNext(t);
                                }
                            });
            rxBusEventArrayMap.put(tag, rxBusEvent);
        }
        return rxBusEvent.disposable;
    }

    /**
     * 取消订阅
     *
     * @param tag 标志
     * @return true 取消成功
     */
    public boolean unregister(@NonNull Object tag) {
        RxBusEvent rxBusEvent = rxBusEventArrayMap.get(tag);
        if (RxUtils.isEmpty(rxBusEvent)) {
            return true;
        }
        Subject<Object> subject = rxBusEvent.subject;
        Disposable disposable = rxBusEvent.disposable;
        if (!disposable.isDisposed()) {
            disposable.dispose();
        }
        if (!subject.hasObservers()) {
            rxBusEventArrayMap.remove(tag);
            return true;
        }
        return false;
    }

    /**
     * 取消所有订阅
     *
     * @return 返回false 证明还有订阅没有被取消,注意内存泄漏...
     */
    public boolean unregisterAll() {
        for (Map.Entry<Object, RxBusEvent> objectRxBusEventEntry : rxBusEventArrayMap.entrySet()) {
            boolean unregister = unregister(objectRxBusEventEntry.getKey());
            if (!unregister) {
                return false;
            }
        }
        return true;
    }
}

File: RxNetWorkLibrary/src/main/java/io/reactivex/network/bus/RxBus.kt
package io.reactivex.network.bus

import android.support.v4.util.ArrayMap
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.network.util.RxUtils
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject

/**
 * by y on 2017/2/22.
 */
class RxBus private constructor() {
    private val rxBusEventArrayMap: ArrayMap<Any, RxBusEvent> = ArrayMap()

    private object RxBusHolder {
        val rxBus = RxBus()
    }

    fun post(obj: Any): Boolean = post(obj, obj)

    fun post(tag: Any, obj: Any): Boolean {
        val rxBusEvent = rxBusEventArrayMap[tag]
        if (!RxUtils.isEmpty(rxBusEvent)) {
            rxBusEvent!!.subject!!.onNext(obj)
            return true
        }
        return false
    }

    /**
     * 接受消息
     *
     * @param tag      标志
     * @param callBack 回调
     */
    fun <T> register(tag: Any,
                     callBack: RxBusCallBack<T>): DisposableObserver<*>? {
        var rxBusEvent = rxBusEventArrayMap[tag]
        if (RxUtils.isEmpty(rxBusEvent)) {
            rxBusEvent = RxBusEvent()
            rxBusEvent.subject = PublishSubject.create<Any>().toSerialized()
            rxBusEvent.disposable = rxBusEvent.subject!!
                    .ofType(callBack.busOfType())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(object : RxBusObserver<T>() {
                        override fun onError(@io.reactivex.annotations.NonNull e: Throwable) {
                            super.onError(e)
                            callBack.onBusError(e)
                        }

                        override fun onNext(@io.reactivex.annotations.NonNull t: T) {
                            super.onNext(t)
                            callBack.onBusNext(t)
                        }
                    })
            rxBusEventArrayMap.put(tag, rxBusEvent)
        }
        return rxBusEvent!!.disposable
    }

    /**
     * 取消订阅
     *
     * @param tag 标志
     * @return true 取消成功
     */
    fun unregister(tag: Any): Boolean {
        val rxBusEvent = rxBusEventArrayMap[tag]
        if (RxUtils.isEmpty(rxBusEvent)) {
            return true
        }
        val subject = rxBusEvent!!.subject
        val disposable = rxBusEvent!!.disposable
        if (!disposable!!.isDisposed) {
            disposable.dispose()
        }
        if (!subject!!.hasObservers()) {
            rxBusEventArrayMap.remove(tag)
            return true
        }
        return false
    }

    /**
     * 取消所有订阅
     *
     * @return 返回false 证明还有订阅没有被取消,注意内存泄漏...
     */
    fun unregisterAll(): Boolean {
        for ((key) in rxBusEventArrayMap) {
            val unregister = unregister(key)
            if (!unregister) {
                return false
            }
        }
        return true
    }

    companion object {

        val instance: RxBus
            get() = RxBusHolder.rxBus
    }
}
